File: lua/atlassian/board/state.lua
local M = {}

---@class BoardNode
---@field issue JiraIssue
---@field children BoardNode[]
---@field expanded boolean
---@field depth number
---@field is_context boolean True if shown only as parent context (dimmed)
---@field urgency string "overdue"|"blocked"|"due_soon"|"stale"|"done"|"normal"

---@class BoardGroup
---@field name string Group header label
---@field nodes BoardNode[]

---@class BoardState
---@field buf number
---@field win number
---@field ns number
---@field issues JiraIssue[]
---@field tree BoardNode[]
---@field groups BoardGroup[]
---@field group_mode string "status"|"assignee"|"type"|"priority"|"due"|"kind"|"cycle"|"none"
---@field line_to_node table<number, BoardNode>
---@field line_count number
---@field filters table
---@field jql string
---@field project string
---@field my_account_id string|nil

---@param issues JiraIssue[]
---@param my_account_id string|nil
---@return BoardNode[]
function M.build_tree(issues, my_account_id)
	local by_key = {}
	for _, issue in ipairs(issues) do
		by_key[issue.key] = issue
	end

	-- Determine which issues directly involve the current user
	local involved = {}
	for _, issue in ipairs(issues) do
		involved[issue.key] = true
	end

	-- Build parent→children mapping
	local children_of = {}
	local has_parent = {}
	for _, issue in ipairs(issues) do
		if issue.parent and by_key[issue.parent] then
			if not children_of[issue.parent] then
				children_of[issue.parent] = {}
			end
			table.insert(children_of[issue.parent], issue.key)
			has_parent[issue.key] = true
		end
	end

	-- Build nodes recursively
	local function build_node(key, depth)
		local issue = by_key[key]
		if not issue then return nil end

		local node = {
			issue = issue,
			children = {},
			expanded = depth == 0,
			depth = depth,
			is_context = not involved[key],
			urgency = M.compute_urgency(issue),
			max_priority_rank = M.priority_rank(issue.priority),
		}

		for _, child_key in ipairs(children_of[key] or {}) do
			local child = build_node(child_key, depth + 1)
			if child then
				table.insert(node.children, child)
				-- Bubble up: parent inherits highest priority (lowest rank number) from children
				if child.max_priority_rank < node.max_priority_rank then
					node.max_priority_rank = child.max_priority_rank
				end
			end
		end

		-- Sort children: blocked/overdue first, then by priority, then by key
		table.sort(node.children, function(a, b)
			local ua = M.urgency_rank(a.urgency)
			local ub = M.urgency_rank(b.urgency)
			if ua ~= ub then return ua < ub end
			if a.max_priority_rank ~= b.max_priority_rank then
				return a.max_priority_rank < b.max_priority_rank
			end
			return a.issue.key < b.issue.key
		end)

		return node
	end

	-- Root nodes: issues with no parent (or parent not in result set)
	local roots = {}
	for _, issue in ipairs(issues) do
		if not has_parent[issue.key] then
			local node = build_node(issue.key, 0)
			if node then
				table.insert(roots, node)
			end
		end
	end

	-- Sort roots: blocked/overdue first, then by priority, then by key
	table.sort(roots, function(a, b)
		local ua = M.urgency_rank(a.urgency)
		local ub = M.urgency_rank(b.urgency)
		if ua ~= ub then return ua < ub end
		if a.max_priority_rank ~= b.max_priority_rank then
			return a.max_priority_rank < b.max_priority_rank
		end
		return a.issue.key < b.issue.key
	end)

	return roots
end

--- Filter a tree for display according to the configured done_filter mode.
--- Counting for the status bar is independent (see M.workable_stats), so hiding
--- done items here does not skew progress.
---   "none"   → show everything
---   "trees"  → hide only fully-done trees (root + all descendants done)
---   "leaves" → also hide done leaf nodes and fully-done subtrees at any level
---@param roots BoardNode[]
---@param mode string|nil
---@return BoardNode[]
function M.filter_for_display(roots, mode)
	mode = mode or "leaves"
	if mode == "none" then
		return roots
	end

	if mode == "trees" then
		local out = {}
		for _, node in ipairs(roots) do
			if not M.is_fully_done(node) then
				out[#out + 1] = node
			end
		end
		return out
	end

	-- "leaves": bottom-up — drop any node that is done and has no (remaining)
	-- children. This removes done leaves and collapses fully-done subtrees.
	local function prune(nodes)
		local out = {}
		for _, node in ipairs(nodes) do
			node.children = prune(node.children)
			local done_leaf = node.urgency == "done" and #node.children == 0
			if not done_leaf then
				out[#out + 1] = node
			end
		end
		return out
	end
	return prune(roots)
end

--- Status-bar stats over the flat fetched issue list (independent of display
--- filtering). "workable" = leaf issues (no children in the fetched set), or the
--- type set named in board.workable_jql when configured.
---@param issues JiraIssue[]
---@return number total, table<string, number> status_counts, number done_count
function M.workable_stats(issues)
	local config = require("jira-interface.config")
	local workable_jql = config.options.board and config.options.board.workable_jql

	local workable_set = nil
	if workable_jql then
		workable_set = {}
		for name in workable_jql:gmatch('"([^"]+)"') do
			workable_set[name:lower()] = true
		end
		if not next(workable_set) then workable_set = nil end
	end

	local by_key, has_child = {}, {}
	for _, issue in ipairs(issues) do by_key[issue.key] = true end
	for _, issue in ipairs(issues) do
		if issue.parent and by_key[issue.parent] then has_child[issue.parent] = true end
	end

	local total, done_count, status_counts = 0, 0, {}
	for _, issue in ipairs(issues) do
		local is_workable
		if workable_set then
			is_workable = workable_set[(issue.type or ""):lower()]
		else
			is_workable = not has_child[issue.key]
		end
		if is_workable then
			total = total + 1
			local s = issue.status or "Unknown"
			status_counts[s] = (status_counts[s] or 0) + 1
			local sl = s:lower()
			if sl:find("done") or sl:find("resolved") or sl:find("closed") then
				done_count = done_count + 1
			end
		end
	end

	return total, status_counts, done_count
end

---@param node BoardNode
---@return boolean
function M.is_fully_done(node)
	if node.urgency ~= "done" then
		return false
	end
	for _, child in ipairs(node.children) do
		if not M.is_fully_done(child) then
			return false
		end
	end
	return true
end

---@param issue JiraIssue
---@return string
function M.compute_urgency(issue)
	local status = (issue.status or ""):lower()
	if status:find("block") or status:find("error") or status:find("fail") then
		return "blocked"
	end
	if status:find("done") or status:find("resolved") or status:find("closed") then
		return "done"
	end

	-- Link-derived blocked: an "is blocked by" link to a non-done issue.
	for _, link in ipairs(issue.links or {}) do
		local is_blocked_by = (link.label or ""):lower():find("blocked by")
			or (link.link_type == "Blocks" and link.direction == "inward")
		if is_blocked_by then
			local ls = (link.issue_status or ""):lower()
			if not (ls:find("done") or ls:find("resolved") or ls:find("closed")) then
				return "blocked"
			end
		end
	end

	if issue.duedate then
		local y, m, d = issue.duedate:match("(%d+)-(%d+)-(%d+)")
		if y then
			local due_ts = os.time({ year = tonumber(y), month = tonumber(m), day = tonumber(d) })
			local now = os.time()
			local days_until = (due_ts - now) / 86400
			if days_until < 0 then
				return "overdue"
			elseif days_until <= 7 then
				return "due_soon"
			end
		end
	end

	if issue.updated then
		local y, m, d = issue.updated:match("(%d+)-(%d+)-(%d+)")
		if y then
			local upd_ts = os.time({ year = tonumber(y), month = tonumber(m), day = tonumber(d) })
			local now = os.time()
			local days_since = (now - upd_ts) / 86400
			if days_since > 14 then
				return "stale"
			end
		end
	end

	return "normal"
end

local urgency_order = {
	overdue = 1,
	blocked = 2,
	due_soon = 3,
	normal = 4,
	stale = 5,
	done = 6,
}

local priority_order = {
	Highest = 1,
	High = 2,
	Medium = 3,
	Low = 4,
	Lowest = 5,
}

---@param priority string|nil
---@return number
function M.priority_rank(priority)
	if not priority then return 99 end
	return priority_order[priority] or 99
end

---@param node BoardNode
---@return string The effective priority name (highest among self + children)
function M.effective_priority(node)
	local rank = node.max_priority_rank or 99
	for name, r in pairs(priority_order) do
		if r == rank then return name end
	end
	return "None"
end

---@param urgency string
---@return number
function M.urgency_rank(urgency)
	return urgency_order[urgency] or 4
end

--- Resolve the configured epic kind for an issue from its labels.
---@param issue JiraIssue
---@return table|nil { name, icon, hl }, string|nil label
function M.epic_kind(issue)
	local config = require("jira-interface.config")
	local kinds = config.options.board and config.options.board.epic_kinds
	if not kinds then return nil end
	for _, label in ipairs(issue.labels or {}) do
		local entry = kinds[label]
		if entry then return entry, label end
	end
	return nil
end

--- Extract a Shape Up cycle identifier from an issue (summary, then fix versions).
---@param issue JiraIssue
---@return string|nil
function M.cycle_id(issue)
	local config = require("jira-interface.config")
	local pattern = config.options.board and config.options.board.cycle_pattern
	if not pattern then return nil end
	local from_summary = issue.summary and issue.summary:match(pattern)
	if from_summary then return from_summary end
	for _, v in ipairs(issue.fix_versions or {}) do
		local m = v:match(pattern)
		if m then return m end
		-- Fix version name may itself be the cycle id (e.g. "SU1/26")
		if v ~= "" then return v end
	end
	return nil
end

--- Evaluate definition-of-readiness for a node.
--- Readiness is a set of issue rules (see atlassian.board.rules) that must all
--- pass for the item's hierarchy level.
---@param node BoardNode
---@return string|nil "ready"|"shaping" (nil when disabled / not applicable / done)
function M.readiness(node)
	local config = require("jira-interface.config")
	local cfg = config.options.board and config.options.board.readiness
	if not cfg or not cfg.enabled then return nil end
	if node.urgency == "done" then return nil end

	local rules = require("atlassian.board.rules")
	rules.load_user_rules()
	local ids = cfg.levels and cfg.levels[node.issue.level]
	local result = rules.all_pass(ids, node)
	if result == nil then return nil end
	return result and "ready" or "shaping"
end

---@param nodes BoardNode[]
---@param mode string
---@return BoardGroup[]
function M.group_nodes(nodes, mode)
	if mode == "none" then
		return { { name = "", nodes = nodes } }
	end

	local groups_map = {}
	local group_order = {}

	for _, node in ipairs(nodes) do
		local key
		if mode == "status" then
			key = node.issue.status or "Unknown"
		elseif mode == "assignee" then
			key = node.issue.assignee or "Unassigned"
		elseif mode == "type" then
			key = node.issue.type or "Unknown"
		elseif mode == "priority" then
			key = M.effective_priority(node)
		elseif mode == "due" then
			local dd = node.issue.duedate
			if not dd then
				key = "No due date"
			else
				local y, m, d = dd:match("(%d+)-(%d+)-(%d+)")
				if y then
					local due_ts = os.time({ year = tonumber(y), month = tonumber(m), day = tonumber(d) })
					local now = os.time()
					local days = (due_ts - now) / 86400
					if days < 0 then
						key = "Overdue"
					elseif days <= 1 then
						key = "Today"
					elseif days <= 7 then
						key = "This week"
					elseif days <= 30 then
						key = "This month"
					else
						key = "Later"
					end
				else
					key = "No due date"
				end
			end
		elseif mode == "kind" then
			local entry = M.epic_kind(node.issue)
			key = entry and entry.name or "Unclassified"
		elseif mode == "cycle" then
			key = M.cycle_id(node.issue) or "No cycle"
		else
			key = "All"
		end

		if not groups_map[key] then
			groups_map[key] = {}
			table.insert(group_order, key)
		end
		table.insert(groups_map[key], node)
	end

	-- Sort groups for modes with natural ordering
	if mode == "priority" then
		local priority_order = { "Highest", "High", "Medium", "Low", "Lowest", "None" }
		local rank = {}
		for i, p in ipairs(priority_order) do rank[p] = i end
		table.sort(group_order, function(a, b)
			return (rank[a] or 99) < (rank[b] or 99)
		end)
	elseif mode == "due" then
		local due_order = { "Overdue", "Today", "This week", "This month", "Later", "No due date" }
		local rank = {}
		for i, d in ipairs(due_order) do rank[d] = i end
		table.sort(group_order, function(a, b)
			return (rank[a] or 99) < (rank[b] or 99)
		end)
	end

	local groups = {}
	for _, name in ipairs(group_order) do
		table.insert(groups, { name = name, nodes = groups_map[name] })
	end
	return groups
end

---@param node BoardNode
---@return number done
---@return number total
function M.subtask_progress(node)
	local done, total = 0, 0
	for _, child in ipairs(node.children) do
		total = total + 1
		local s = (child.issue.status or ""):lower()
		if s:find("done") or s:find("resolved") or s:find("closed") then
			done = done + 1
		end
	end
	return done, total
end

---@param done number
---@param total number
---@param width number
---@return string
function M.progress_bar(done, total, width)
	if total == 0 then return "" end
	local filled = math.floor((done / total) * width)
	return string.rep("█", filled) .. string.rep("░", width - filled)
end

return M

File: core/board/state.go
package board

import (
	"regexp"
	"sort"
	"strings"
	"time"
)

var priorityOrder = map[string]int{"Highest": 1, "High": 2, "Medium": 3, "Low": 4, "Lowest": 5}
var urgencyOrder = map[string]int{"overdue": 1, "blocked": 2, "due_soon": 3, "normal": 4, "stale": 5, "done": 6}

func priorityRank(p string) int {
	if p == "" {
		return 99
	}
	if r, ok := priorityOrder[p]; ok {
		return r
	}
	return 99
}

func urgencyRank(u string) int {
	if r, ok := urgencyOrder[u]; ok {
		return r
	}
	return 4
}

func statusDone(s string) bool {
	s = strings.ToLower(s)
	return strings.Contains(s, "done") || strings.Contains(s, "resolved") || strings.Contains(s, "closed")
}

// computeUrgency mirrors state.lua compute_urgency.
func computeUrgency(issue *Issue) string {
	status := strings.ToLower(issue.Status)
	if strings.Contains(status, "block") || strings.Contains(status, "error") || strings.Contains(status, "fail") {
		return "blocked"
	}
	if statusDone(issue.Status) {
		return "done"
	}
	for _, link := range issue.Links {
		blockedBy := strings.Contains(strings.ToLower(link.Label), "blocked by") ||
			(link.LinkType == "Blocks" && link.Direction == "inward")
		if blockedBy && !statusDone(link.IssueStatus) {
			return "blocked"
		}
	}
	if days, ok := daysFromNow(issue.Duedate); ok {
		if days < 0 {
			return "overdue"
		} else if days <= 7 {
			return "due_soon"
		}
	}
	if d, ok := daysSince(issue.Updated); ok && d > 14 {
		return "stale"
	}
	return "normal"
}

var dateRe = regexp.MustCompile(`(\d+)-(\d+)-(\d+)`)

func parseDate(s string) (time.Time, bool) {
	m := dateRe.FindStringSubmatch(s)
	if m == nil {
		return time.Time{}, false
	}
	t, err := time.Parse("2006-1-2", m[1]+"-"+m[2]+"-"+m[3])
	if err != nil {
		return time.Time{}, false
	}
	return t, true
}

func daysFromNow(s string) (float64, bool) {
	t, ok := parseDate(s)
	if !ok {
		return 0, false
	}
	return time.Until(t).Hours() / 24, true
}

func daysSince(s string) (float64, bool) {
	t, ok := parseDate(s)
	if !ok {
		return 0, false
	}
	return time.Since(t).Hours() / 24, true
}

func effectivePriority(n *Node) string {
	rank := n.MaxPriorityRank
	for name, r := range priorityOrder {
		if r == rank {
			return name
		}
	}
	return "None"
}

// BuildTree mirrors state.lua build_tree (returns the full sorted tree; display
// filtering is separate).
func BuildTree(issues []*Issue) []*Node {
	byKey := map[string]*Issue{}
	involved := map[string]bool{}
	for _, is := range issues {
		byKey[is.Key] = is
		involved[is.Key] = true
	}

	childrenOf := map[string][]string{}
	hasParent := map[string]bool{}
	for _, is := range issues {
		if is.Parent != "" && byKey[is.Parent] != nil {
			childrenOf[is.Parent] = append(childrenOf[is.Parent], is.Key)
			hasParent[is.Key] = true
		}
	}

	var build func(key string, depth int) *Node
	build = func(key string, depth int) *Node {
		issue := byKey[key]
		if issue == nil {
			return nil
		}
		n := &Node{
			Issue:           issue,
			Expanded:        depth == 0,
			Depth:           depth,
			IsContext:       !involved[key],
			Urgency:         computeUrgency(issue),
			MaxPriorityRank: priorityRank(issue.Priority),
		}
		for _, ck := range childrenOf[key] {
			child := build(ck, depth+1)
			if child != nil {
				n.Children = append(n.Children, child)
				if child.MaxPriorityRank < n.MaxPriorityRank {
					n.MaxPriorityRank = child.MaxPriorityRank
				}
			}
		}
		sortNodes(n.Children)
		return n
	}

	var roots []*Node
	for _, is := range issues {
		if !hasParent[is.Key] {
			if n := build(is.Key, 0); n != nil {
				roots = append(roots, n)
			}
		}
	}
	sortNodes(roots)
	return roots
}

func sortNodes(nodes []*Node) {
	sort.Slice(nodes, func(i, j int) bool {
		a, b := nodes[i], nodes[j]
		ua, ub := urgencyRank(a.Urgency), urgencyRank(b.Urgency)
		if ua != ub {
			return ua < ub
		}
		if a.MaxPriorityRank != b.MaxPriorityRank {
			return a.MaxPriorityRank < b.MaxPriorityRank
		}
		return a.Issue.Key < b.Issue.Key
	})
}

func isFullyDone(n *Node) bool {
	if n.Urgency != "done" {
		return false
	}
	for _, c := range n.Children {
		if !isFullyDone(c) {
			return false
		}
	}
	return true
}

// FilterForDisplay mirrors state.lua filter_for_display.
func FilterForDisplay(roots []*Node, mode string) []*Node {
	if mode == "" {
		mode = "leaves"
	}
	switch mode {
	case "none":
		return roots
	case "trees":
		var out []*Node
		for _, n := range roots {
			if !isFullyDone(n) {
				out = append(out, n)
			}
		}
		return out
	default: // leaves
		var prune func(nodes []*Node) []*Node
		prune = func(nodes []*Node) []*Node {
			var out []*Node
			for _, n := range nodes {
				n.Children = prune(n.Children)
				if !(n.Urgency == "done" && len(n.Children) == 0) {
					out = append(out, n)
				}
			}
			return out
		}
		return prune(roots)
	}
}

var cycleRe = regexp.MustCompile(CyclePattern)

// EpicKindOf returns the configured kind for an issue's labels, or nil.
func EpicKindOf(issue *Issue) *EpicKind {
	for _, label := range issue.Labels {
		if k, ok := EpicKinds[label]; ok {
			return &k
		}
	}
	return nil
}

// CycleID extracts a cycle id from summary then fix versions.
func CycleID(issue *Issue) string {
	if m := cycleRe.FindString(issue.Summary); m != "" {
		return m
	}
	for _, v := range issue.FixVersions {
		if m := cycleRe.FindString(v); m != "" {
			return m
		}
		if v != "" {
			return v
		}
	}
	return ""
}

// Readiness returns "ready"|"shaping"|"" (empty = no glyph).
func Readiness(n *Node) string {
	if !ReadinessEnabled || n.Urgency == "done" {
		return ""
	}
	ids := ReadinessLevels[n.Issue.Level]
	if len(ids) == 0 {
		return ""
	}
	for _, id := range ids {
		if !readinessRule(id, n) {
			return "shaping"
		}
	}
	return "ready"
}

func readinessRule(id string, n *Node) bool {
	switch id {
	case "description":
		return n.Issue.Description != ""
	case "acceptance_criteria":
		return n.Issue.AcceptanceCriteria != ""
	case "child_task":
		return len(n.Children) > 0
	case "fix_version":
		return len(n.Issue.FixVersions) > 0
	case "epic_link":
		return n.Issue.Parent != ""
	case "assignee":
		return n.Issue.Assignee != ""
	}
	return true // unknown rule id: skip (never fails)
}

// SubtaskProgress returns done/total over direct children.
func SubtaskProgress(n *Node) (done, total int) {
	for _, c := range n.Children {
		total++
		if statusDone(c.Issue.Status) {
			done++
		}
	}
	return
}

// WorkableStats mirrors state.lua workable_stats (workable = leaf issues).
func WorkableStats(issues []*Issue) (total int, statusCounts map[string]int, doneCount int) {
	statusCounts = map[string]int{}
	byKey := map[string]bool{}
	hasChild := map[string]bool{}
	for _, is := range issues {
		byKey[is.Key] = true
	}
	for _, is := range issues {
		if is.Parent != "" && byKey[is.Parent] {
			hasChild[is.Parent] = true
		}
	}
	for _, is := range issues {
		if hasChild[is.Key] {
			continue
		}
		total++
		s := is.Status
		if s == "" {
			s = "Unknown"
		}
		statusCounts[s]++
		if statusDone(is.Status) {
			doneCount++
		}
	}
	return
}

// GroupNodes mirrors state.lua group_nodes.
func GroupNodes(nodes []*Node, mode string) []Group {
	if mode == "none" {
		return []Group{{Name: "", Nodes: nodes}}
	}
	order := []string{}
	m := map[string][]*Node{}
	keyOf := func(n *Node) string {
		switch mode {
		case "status":
			if n.Issue.Status != "" {
				return n.Issue.Status
			}
			return "Unknown"
		case "assignee":
			if n.Issue.Assignee != "" {
				return n.Issue.Assignee
			}
			return "Unassigned"
		case "type":
			if n.Issue.Type != "" {
				return n.Issue.Type
			}
			return "Unknown"
		case "priority":
			return effectivePriority(n)
		case "kind":
			if k := EpicKindOf(n.Issue); k != nil {
				return k.Name
			}
			return "Unclassified"
		case "cycle":
			if c := CycleID(n.Issue); c != "" {
				return c
			}
			return "No cycle"
		}
		return "All"
	}
	for _, n := range nodes {
		k := keyOf(n)
		if _, ok := m[k]; !ok {
			order = append(order, k)
		}
		m[k] = append(m[k], n)
	}
	if mode == "priority" {
		rank := map[string]int{"Highest": 1, "High": 2, "Medium": 3, "Low": 4, "Lowest": 5, "None": 6}
		sort.SliceStable(order, func(i, j int) bool { return at(rank, order[i]) < at(rank, order[j]) })
	}
	out := make([]Group, 0, len(order))
	for _, name := range order {
		out = append(out, Group{Name: name, Nodes: m[name]})
	}
	return out
}

func at(m map[string]int, k string) int {
	if v, ok := m[k]; ok {
		return v
	}
	return 99
}

// BuildState runs the full pipeline: tree → display filter → grouping.
func BuildState(in *BoardInput) *State {
	tree := FilterForDisplay(BuildTree(in.Issues), "leaves")
	return &State{
		Issues:  in.Issues,
		Tree:    tree,
		Groups:  GroupNodes(tree, in.Group),
		Project: in.Project,
	}
}
